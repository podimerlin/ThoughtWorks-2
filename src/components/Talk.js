import React, { Component } from 'react'

class Talk extends Component {
    constructor(props) {
        super(props);
        var initDate = new Date();
        initDate.setHours(9,0,0,0);
        this.state = { ...props, time: initDate };                
    }

    // add minutes to 9am and display a 12 hour time string
    changeStringToTime(time, mins) {
      mins = typeof this.props.length === 'undefined' ? mins : mins - this.props.length;
      time = new Date(time.getTime() + mins*60000);

      return (
        time.toLocaleString('en-US', { hour: 'numeric',minute:'numeric', hour12: true })
      );
    }

    render() {
        return(
            <div> {this.changeStringToTime(this.state.time, this.props.spentTime)} {this.state.title} {this.state.lengthtext} </div>
        )
    }   
}

export default Talk;