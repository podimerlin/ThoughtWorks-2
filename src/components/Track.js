import React, { Component } from 'react';
import Morning from './Morning';
import Afternoon from './Afternoon';
import Talk from './Talk';


class Track extends Component {

  constructor(props) {
    super(props);

    this.length = 0;
    //minimum is 360, max is 420
    this.morningTalks = [];
    this.afternoonTalks = [];

    //split for Morning and Afternoon
    this.splitTracks = this.splitTracks.bind(this);
    this.splitTracks();

    this.state = {
      morningTalks : this.morningTalks,
      afternoonTalks : this.afternoonTalks
    }
  }

  render() {
    return( 
      <div className="track">
        Track: {this.props.id}
        <Morning key="morning" talks={this.state.morningTalks}/>
        <Talk title='Lunch' className="lunch" spentTime={180}/>
        <Afternoon key="afternoon" talks={this.state.afternoonTalks}/>
      </div>
    );
  }

  splitTracks() {
    //morning only fit max 180
    this.props.talks.forEach((talk) => {      
      if(this.length + talk.props.length <= 180 ) {
        this.length += talk.props.length;
        this.morningTalks.push(talk);
      }
    },this);

    this.afternoonTalks = this.props.talks;

    this.morningTalks.forEach((talk) => {
      const t2 = this.afternoonTalks.filter(talk2 => talk2.key !== talk.key);      
      this.afternoonTalks = t2;
    });    
    
  }

}

export default Track;
