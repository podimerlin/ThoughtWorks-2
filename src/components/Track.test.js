import React from 'react';
import { shallow } from 'enzyme';
import Track from './Track';
import Talk from './Talk';

describe('Track', () => {    
  const talkid = 1;
  const props = {talks: [ <Talk title={'title'} length={10} lengthtext={'lengthtext'} /> ], length: 0};
  const track = shallow(<Track {...props} />);
  
  it('renders correctly', () => {
      expect(track).toMatchSnapshot();
  });

  it('creates Morning and Afternoon components', () => {
    expect(track.find('Morning').exists()).toBe(true);
    expect(track.find('Afternoon').exists()).toBe(true);
  });

  it('create a lunch talk component', () => {
    expect(track.find('Talk.lunch').exists()).toBe(true);
  })

})
