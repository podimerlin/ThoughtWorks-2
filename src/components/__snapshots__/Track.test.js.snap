// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Track renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Track
    length={0}
    talks={
        Array [
            <Talk
              length={10}
              lengthtext="lengthtext"
              title="title"
        />,
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        "Track: ",
        undefined,
        <Morning
          talks={
                    Array [
                              <Talk
                                length={10}
                                lengthtext="lengthtext"
                                title="title"
                    />,
                            ]
          }
/>,
        <Talk
          className="lunch"
          spentTime={180}
          title="Lunch"
/>,
        <Afternoon
          talks={Array []}
/>,
      ],
      "className": "track",
    },
    "ref": null,
    "rendered": Array [
      "Track: ",
      undefined,
      Object {
        "instance": null,
        "key": "morning",
        "nodeType": "class",
        "props": Object {
          "talks": Array [
            <Talk
              length={10}
              lengthtext="lengthtext"
              title="title"
/>,
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "lunch",
          "spentTime": 180,
          "title": "Lunch",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "afternoon",
        "nodeType": "class",
        "props": Object {
          "talks": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Track: ",
          undefined,
          <Morning
            talks={
                        Array [
                                    <Talk
                                      length={10}
                                      lengthtext="lengthtext"
                                      title="title"
                        />,
                                  ]
            }
/>,
          <Talk
            className="lunch"
            spentTime={180}
            title="Lunch"
/>,
          <Afternoon
            talks={Array []}
/>,
        ],
        "className": "track",
      },
      "ref": null,
      "rendered": Array [
        "Track: ",
        undefined,
        Object {
          "instance": null,
          "key": "morning",
          "nodeType": "class",
          "props": Object {
            "talks": Array [
              <Talk
                length={10}
                lengthtext="lengthtext"
                title="title"
/>,
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "lunch",
            "spentTime": 180,
            "title": "Lunch",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "afternoon",
          "nodeType": "class",
          "props": Object {
            "talks": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
